File: Data Structures & Algorithms/count-number-of-islands/submission-3.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        ROWS, COLS = len(grid), len(grid[0])
        lands = 0
        def dfs (i,j):
            if i<0 or j<0 or i>=ROWS or j>=COLS or grid[i][j]=="0":
                return
            grid[i][j]="0"
            for x,y in directions:
                dfs(i+x,j+y)

        for i in range(ROWS):
            for j in range(COLS):
                if grid[i][j]=="1":
                    lands+=1
                    dfs(i,j)
        return lands

File: Data Structures & Algorithms/max-area-of-island/submission-3.cpp
class Solution {
    int ROWS, COLS;
    int directions[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    int dfs(int i, int j, vector<vector<int>>& grid) {
        if (i < 0 || j < 0 || i >= ROWS || j >= COLS || grid[i][j] == 0) {
            return 0;
        }
        grid[i][j] = 0;
        int curr = 0;
        for (auto& dir : directions) {
            curr += dfs(i + dir[0], j + dir[1], grid);
        }
        return 1+curr;
    }

   public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        ROWS = grid.size();
        COLS = grid[0].size();
        int res = 0;
        for (int i = 0; i < ROWS; i++) {
            for (int j = 0; j < COLS; j++) {
                if (grid[i][j] == 1) {
                    res = max(res, dfs(i, j, grid));
                }
            }
        }
        return res;
    }
};
